docs/developer-docs/smart-contracts/advanced-features/async-code.mdx

```diff
@@ -1,90 +1,94 @@
+
+Furthermore, canister methods that use inter-canister calls are **not atomic**. A failure somewhere in the method might not roll back all the changes that the method performed. Under the hood, such methods use the (low-level Wasm API)[### Low-level API] calls, and are translated into multiple *message handler Wasm functions*: an initial handler function that handles the method call itself (corresponding to `do_work` above), and other handler functions that serve as callbacks to handle the responses for any inter-canister calls that have been issued (`do_more_work` in the example above). The execution of each message handler is atomic on its own; an error (more precisely, a Wasm trap) rolls back only the changes performed by the current message handler, but does not affect the changes made by the other handlers.
 
-When `baz().await` is called a future is created to represent its result. The result can actually be immediately available since there's only
-local computation (i.e. no inter-canister call involved) happening in `baz`. So, the future can return a result directly when polled and message
-execution can still proceed up to the point where the call to canister B is made (similar to the example from the previous section).
+For these reasons, to write correct code in the presence of inter-canister calls you should understand how the CDK for your language of choice desugars the `async/await` syntax into Wasm code, and in particular where the message handler boundaries are, as these are both "commit points" for state changes and potential interleaving points for concurrent executions.
 
-The above behavior is one possibility depending on how the language runtime handles the `async/await` pattern. Rust follows this approach.
-In contrast, Motoko takes another approach where every async/await call will be converted to an inter-canister call, more specifically a self-call
-in case a local function is called. This means that in Motoko `foo` would be split across 3 message executions -- the first would be until `baz` is
-called, the second until canister B is called and the final one until the end of the function body.
+A [separate document](https://internetcomputer.org/docs/current/references/message-execution-properties) contains more details on interleaving/commit points and message execution properties. Refer to the [security best practices](/docs/current/developer-docs/security/security-best-practices/inter-canister-calls) for advice on how to handle concurrency and state rollback issues correctly. Finally, refer to the CDK documentation for your language to learn more about desugaring of `async/await` and interleaving/commit points.
```

Comment:
Should we also point to the security best practices for best-effort responses?

Reply:
We have a single security best practices doc for all types of inter-canister calls (as most points are shared between both call types), and we already link to it above.